brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+## Metadata (uns)
+
+All SOMA objects may be annotated with a small amount of metadata, in the form of a simple key/value map:
+
+- `uns` (map[string, simple]) - string-keyed mapping to any **simple** type (eg, int32, string, etc).
+
+Only `simple` types are supported in the metadata values. The metadata lifecycle is the same as its containing object, eg, it will be deleted when the containing object is deleted.
+
+> ⚠️ **Issue** - `uns` - do we need more complex typing for uns values, and if so, how to represent capture what can is required to be supported? (eg, anything Arrow can represent? a subset?)
+>
+> **Proposal**: restrict `uns` to a mapping of string keys and `simple` values. Extend to complex types in the future as we have concrete use cases which warrant the complexity.
+
+> ⚠️ **Issue** - the name `uns` is not really all that suggestive. **Proposal:** rename `uns` to `metadata`, eg, `print(soma.metadata['version'])`
+
+## SOMACollection
+
+SOMACollection is an unordered, `string`-keyed map of values of type SOMADataFrame, SOMAMatrix, SOMACollection or SOMA. SOMACollection objects are typically persistent, and may be nested to any level (eg, SOMACollection may contain other SOMACollection objects). Keys in the map are unique and singular (no duplicates, ie, it is not a multi-map).
+
+## SOMADataFrame
+
+SOMADataFrame is a 1-D multi-column table of monomorphic arrays, all of equal length -- ie, a "dataframe". The SOMADataFrame has a user-defined schema, which includes:
+
+- a table schema, expressed as an `Arrow.RecordBatch` type, defining number of columns, and their respective names and types
+- the index column name
+
+A SOMADataFrame must contain at least a single column which is the named index column. The index column _must_ be of `string` type. Index values are non-null and must be unique (ie, no duplicates are allowed) within the object. The index is sparse and unordered, ie, supported indexing is by ID value only, with no indexing order or positional (offset-based) indexing.
+
+Most language-specific implementations will present SOMADataFrame in more convenient forms, such as Python `pandas.DataFrame`, R `data.frame` or other environment-specific data structures. Specific implementations of the SOMA API may extend this type with convertors to/from the underlying data structure, and enforce indexing conventions.
+
+> ⚠️ **Issue** - this spec requires that index column is a `string` type. Most underlying "engines" will support more flexibility. Do we want to enable that flexibility, or stick with requiring string for simplicity/commonality?
```

Comment:
i prefer we just say string & hold to that ...

Reply:
obsolete issue given latest discussions.